File: frontend/frontend/formfilling/libs/source.js

/* 
This is the data source module for the formfilling project. Currently it only supports excel file, but it will be extended to support database in the future.
1. Read excel file and convert to json. The info sheet should start with "info-" and the table sheet should start with "table-", which all follows the naming convention of the source 3.0
2. TODO: read all schema of the source model

*/

const XLSX = require('xlsx');

class Excel {

    constructor(filename) {
        this.filename = filename;
    }

    getDateString(dateNum) {
        return XLSX.SSF.format('yyyy-mm-dd', dateNum);
    }

    json() {
        let excel_data = {}
        const workbook = XLSX.readFile(this.filename);
        const sheetNames = workbook.SheetNames;

        for (let sheetName of sheetNames) {
            const worksheet = workbook.Sheets[sheetName];
            const data = XLSX.utils.sheet_to_row_object_array(worksheet);
            if (sheetName.startsWith("info-")) {
                // info name as key . Get the info name from sheet name by removing "info-"
                const info_name = sheetName.substring(5);
                excel_data[info_name] = {};
                // loop data starting from 2 to skip the header
                for (let i = 2; i < data.length; i++) {
                    // get the key and value from the row
                    const row_values = Object.values(data[i]);
                    const [key, , , value] = row_values;
                    // Check if the value is a date
                    if (key.includes("date") || key.includes("dob")) {
                        excel_data[info_name][key] = this.getDateString(value);
                    } else {
                        excel_data[info_name][key] = value;
                    }
                }
            } else if (sheetName.startsWith("table-")) {
                // table name as key . Get the table name from sheet name by removing "table-"
                const table_name = sheetName.substring(6);
                excel_data[table_name] = [];
                // get the table variables 
                const variables = Object.values(data[1]);
                // get indexes of the variables including "date"
                const date_indexes = variables.map((v, i) => v.includes("date") ? i : null).filter(v => v !== null);

                // loop data starting from 4 to skip the header
                for (let i = 4; i < data.length + 1; i++) {
                    let row = {};
                    for (let j = 0; j < variables.length; j++) {
                        // Check if the value is a date
                        const cell = worksheet[XLSX.utils.encode_cell({ r: i, c: j })];

                        let column_value;
                        if (cell && (cell.v === 0 || cell.v === '0')) {
                            column_value = cell.v;
                        } else {
                            column_value = (cell && cell.v) ? cell.v : undefined;
                        }

                        if (date_indexes.includes(j)) {
                            if (column_value === undefined || column_value === null || column_value === "") {
                                row[variables[j]] = null;
                            } else {
                                row[variables[j]] = this.getDateString(column_value);
                            }
                        } else {
                            row[variables[j]] = column_value;
                        }
                    }
                    // check special tables for contact, eradddress, 
                    switch (table_name) {
                        case "contact":
                            if (row.first_name) excel_data[table_name].push(row); // only push the row if the first name is not empty
                            break;
                        case "eraddress":
                        case "address":
                            if (row.street_name) excel_data[table_name].push(row); // only push the row if the street name is not empty
                            break;
                        case "phone":
                        case "personid":
                            if (row.number) excel_data[table_name].push(row); // only push the row if the number is not empty
                            break;
                        default:
                            excel_data[table_name].push(row);
                            break;
                    }
                }
            }
        }

        // console.log(JSON.stringify(excel_data, null, 2));
        return excel_data;
    }
}

// 2. Get data from database. TBD

// class Database {}


module.exports = {
    Excel
}

File: client/client/app/util/source.go
package util

import (
	"encoding/json"
	"time"

	"github.com/tealeg/xlsx"
)

type Excel struct {
	filename string
}

func NewExcel(filename string) *Excel {
	return &Excel{
		filename: filename,
	}
}

func (e *Excel) getDateString(dateNum float64) string {
	t := time.Unix(int64(dateNum), 0)
	return t.Format("2006-01-02")
}

func (e *Excel) Json() ([]byte, error) {
	excelData := make(map[string]interface{})
	workbook, err := xlsx.OpenFile(e.filename)
	if err != nil {
		return []byte{}, err
	}

	for _, sheet := range workbook.Sheets {
		sheetName := sheet.Name
		data := sheet.Rows

		if sheetName[:5] == "info-" {
			infoName := sheetName[5:]
			excelData[infoName] = make(map[string]interface{})

			for i := 3; i < len(data); i++ {
				row := data[i].Cells
				key := row[0].Value
				value := row[3].Value

				if key != "" {
					if keyContainsDate(key) {
						dateValue, _ := row[3].Float()
						excelData[infoName].(map[string]interface{})[key] = e.getDateString(dateValue)
					} else {
						excelData[infoName].(map[string]interface{})[key] = value
					}
				}
			}
		} else if sheetName[:6] == "table-" {
			tableName := sheetName[6:]
			excelData[tableName] = make([]map[string]interface{}, 0)
			variables := data[2].Cells

			dateIndexes := make([]int, 0)
			for i, variable := range variables {
				if keyContainsDate(variable.Value) {
					dateIndexes = append(dateIndexes, i)
				}
			}

			for i := 4; i < len(data); i++ {
				row := data[i].Cells
				tableRow := make(map[string]interface{})
				isEmptyRow := true

				for j, variable := range variables {
					cell := row[j]
					columnValue := cell.Value

					if cell.Type() == xlsx.CellTypeNumeric && cell.Value == "0" {
						columnValue = cell.Value
					}

					if contains(dateIndexes, j) {
						if columnValue == "" {
							tableRow[variable.Value] = nil
						} else {
							dateValue, _ := cell.Float()
							tableRow[variable.Value] = e.getDateString(dateValue)
						}
					} else {
						tableRow[variable.Value] = columnValue
					}
					if columnValue != "" {
						isEmptyRow = false
					}
				}
				if !isEmptyRow {
					switch tableName {
					case "contact":
						if tableRow["first_name"].(string) != "" {
							excelData[tableName] = append(excelData[tableName].([]map[string]interface{}), tableRow)
						}
					case "eraddress", "address":
						if tableRow["street_name"].(string) != "" {
							excelData[tableName] = append(excelData[tableName].([]map[string]interface{}), tableRow)
						}
					case "phone", "personid":
						if tableRow["number"].(string) != "" {
							excelData[tableName] = append(excelData[tableName].([]map[string]interface{}), tableRow)
						}
					default:
						excelData[tableName] = append(excelData[tableName].([]map[string]interface{}), tableRow)
					}
				}

			}
		}
	}
	jsonData, err := json.MarshalIndent(excelData, "", "  ")
	if err != nil {
		return []byte{}, err
	}

	return jsonData, nil
}

func keyContainsDate(key string) bool {
	return keyContainsSubstring(key, "date") || keyContainsSubstring(key, "dob")
}

func keyContainsSubstring(key, substr string) bool {
	return len(key) >= len(substr) && key[:len(substr)] == substr
}

func contains(arr []int, val int) bool {
	for _, v := range arr {
		if v == val {
			return true
		}
	}
	return false
}

// func do() {
// 	e := NewExcel("/Users/jacky/desktop/pgp/yx.xlsx")
// 	jsonData, err := e.Json()
// 	if err != nil {
// 		fmt.Println(err)
// 		return
// 	}
// 	fmt.Println(jsonData)
// }

// func main() {
// 	do()
// }
